### PR TITLE
Add support to MaxStalenessSeconds in ReadPreference

### PR DESCRIPTION
**Pull request by wpjunior**
from #270

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -679,6 +689,9 @@ type ReadPreference struct {
 	// Mode determines the consistency of results. See Session.SetMode.
 	Mode Mode
 
+	// MaxStalenessSeconds specify a maximum replication lag, or “staleness” in seconds, for reads from secondaries.
```
Comment (szank) [inline session.go:707]: Could you please add a comment stating that this option is supported in mongo >= 3.4 only ?

File: session.go [hunk 2]
```diff
@@ -390,6 +391,11 @@ func ParseURL(url string) (*DialInfo, error) {
 			if err != nil {
 				return nil, errors.New("bad value for maxPoolSize: " + opt.value)
 			}
+		case "maxStalenessSeconds":
```
Comment (eminano) [inline session.go:399]: According to the documentation maxStalenessSeconds is not compatible with mode primary, so maybe you could add that restriction to the existing check below (line 460).

File: session.go [hunk 3]
```diff
@@ -2190,6 +2204,22 @@ func (s *Session) SetPoolTimeout(timeout time.Duration) {
...
 
+// SetMaxStalenessSeconds set the maximum of seconds of replication lag from secondaries
+//
+// Relevant documentation:
+//
+//   https://docs.mongodb.com/manual/core/read-preference/#maxstalenessseconds
+//
+func (s *Session) SetMaxStalenessSeconds(seconds int) {
+	s.m.Lock()
+	if seconds > -1 {
```
Comment (eminano) [inline session.go]: The documentation states you must specify a maxStalenessSeconds value of 90 seconds or longer (smaller values will raise an error). Since you're doing a validity check here it might be worth checking that as well.

File: session.go [hunk 4]
```diff
@@ -2190,6 +2204,22 @@ func (s *Session) SetPoolTimeout(timeout time.Duration) {
...
 }
 
+// SetMaxStalenessSeconds set the maximum of seconds of replication lag from secondaries
+//
+// Relevant documentation:
+//
+//   https://docs.mongodb.com/manual/core/read-preference/#maxstalenessseconds
+//
+func (s *Session) SetMaxStalenessSeconds(seconds int) {
+	s.m.Lock()
```
Comment (domodwyer) [inline session.go:2231]: It's usually nicer for users if a function returns an error rather than panicking - it's easier to handle and a little less surprising (it's documented as part of the API that there can be an error, rather than just calling something and it dying).

File: session.go [hunk 5]
```diff
@@ -390,6 +391,11 @@ func ParseURL(url string) (*DialInfo, error) {
 			if err != nil {
 				return nil, errors.New("bad value for maxPoolSize: " + opt.value)
 			}
+		case "maxStalenessSeconds":
+			maxStalenessSeconds, err = strconv.Atoi(opt.value)
+			if err != nil {
```
Comment (domodwyer) [inline session.go:402]: I think this would need the `maxStalenessSeconds  >= 0` check to compliment the `SetMaxStalenessSeconds()` behaviour?

File: session.go [hunk 6]
```diff
@@ -2190,6 +2204,22 @@ func (s *Session) SetPoolTimeout(timeout time.Duration) {
 	s.m.Unlock()
 }
 
+// SetMaxStalenessSeconds set the maximum of seconds of replication lag from secondaries
+//
+// Relevant documentation:
```
Comment (domodwyer) [inline session.go:2226]: It might be worth including that if `seconds > 90` it's currently expected to receive an error from mongodb:

> You must specify a maxStalenessSeconds value of 90 seconds or longer: specifying a smaller maxStalenessSeconds value will raise an error.

File: session.go [hunk 7]
```diff
@@ -679,6 +689,9 @@ type ReadPreference struct {
 	// Mode determines the consistency of results. See Session.SetMode.
 	Mode Mode
 
+	// MaxStalenessSeconds specify a maximum replication lag, or “staleness” in seconds, for reads from secondaries.
+	MaxStalenessSeconds int
```
Comment (domodwyer) [inline session.go:708]: Is this missing an `omitempty` tag? I imagine you don't want to send `maxStalenessSeconds: 0` with every request if it was not set by the user.

**Review by domodwyer — CHANGES_REQUESTED** [review]
Is there any way to add functional tests for this, beyond the config parsing? Seems like there might be a few edge cases.

**Review by eminano — CHANGES_REQUESTED** [review]

**Review by szank — CHANGES_REQUESTED** [review]

File: session.go (hunk 2 above)
Comment (szank) [reply session.go:399]: Also, could you please add description of this parameter to the Dial method documentation (i.e. https://godoc.org/github.com/globalsign/mgo#Dial)

**Review by szank — CHANGES_REQUESTED** [review]

File: session.go (hunk 7 above)
Comment (wpjunior) [reply session.go:708]: hi @domodwyer, the wire information is sent by socket.go(line: 131), I suppose that tag is not necessary here, thanks.

File: session.go (hunk 5 above)
Comment (wpjunior) [reply session.go:402]: thanks @domodwyer, I've added you recommendation.

File: session.go (hunk 2 above)
Comment (wpjunior) [reply session.go:399]: thanks @eminano and @szank, could you see my commits updates ?, I committed some changes.

**Review by szank — APPROVED** [review]

**Review by maitesin — APPROVED** [review]

**Review by eminano — COMMENTED** [review]
As @domodwyer mentioned previously, is it possible to add functional tests for this?

**Comment by wpjunior** [comment]
Hi @eminano, would you check the possibility to approve this MR ?, this scenario is too hard to create an integration test. 

Thanks =)

**Comment by wangke1020** [comment]
It seems the changes only pass a ReadPreference params to server and did not select proper servers, so at my guess the driver will not work with mongo replset but sharding cluster. 
According to https://github.com/mongodb/specifications/blob/master/source/max-staleness/max-staleness.rst#client, server selection should be implemented in client side.
